_posts/2022-09-08-storage-write-removal-before-conditional-termination.md

```diff
@@ -0,0 +1,168 @@
+If a function call is performed between (1) and (3), the optimizer has to consider the control flow behaviour of the function call, e.g. whether control flow may continue beyond the call, whether the called function always reverts or always successfully terminates. However, in the case that the control flow conditionally can continue after the call to a function, but the function call may also terminate using the ``return(...)`` or ``stop()`` statements (see examples below), the optimizer incorrectly still behaved, as if the control flow *always* continued after the function.
+
+Hence the write in (1) could be removed, even though (2) can in fact terminate and is supposed to retain the storage write in (1).
+
+So on the Yul level the issue manifests if the following sequence occurs within a Yul block during optimization:
+
+1. A storage write occurs (e.g. ``sstore(0, 1)``, but the write may also occur within complex control flow like a single ``switch`` case).
+2. A user-defined function is called and this function both has a control flow path that terminates the transaction and a control flow path out of the function.
+3. Any continuing control flow path does either of the following:
+    - It overwrites the storage slot written in (1), s.t. the storage write in (1) becomes redundant for this path (e.g. ``sstore(0, 2)``).
+    - The control flow unconditionally reverts (e.g. ``revert(0, 0)``)
+
+In this situation, the optimizer will treat the storage write in (1) as redundant and remove it, despite the fact that the called Yul function may successfully terminate the EVM call, in which case this storage write is in fact not redundant at all.
+
+The following is a full example of an affected Yul snippet:
+
+```yul
+{
+	function f() {
+		if gt(calldatasize(), 4) { leave }
+		return(0, 0)
+	}
+	sstore(0, 1) // This sstore will incorrectly be removed.
+	f()
+	sstore(0, 2)
+}
+```
+
+In the absence of the bug, the first ``sstore`` would have been be marked as required by the optimizer upon encountering the call to ``f()``, i.e. it should not be considered for removal regardless of what happens afterwards, since the control flow path of ``f()`` reaching the ``return(0, 0)`` already relies on the ``sstore`` happening.
+However, the optimizer failed to mark it as such, marking the first ``sstore`` as redundant instead once it sees the second ``sstore`` overwriting the slot.
+
+The issue only occurs in the presence of a call to a conditionally successfully terminating Yul function that does not end up being inlined by the preceding optimizer steps.
+The following very similarly looking snippet is not affected, since the optimizer will inline the call to ``f()`` before performing
+the buggy optimization step:
+
+```yul
+{
+	function f() {
+		if iszero(lt(calldatasize(), 4)) { return(0,0) }
+	}
+	sstore(0, 1)
+	f()
+	sstore(0, 2)
+}
+```
+
+In this case, ``f()`` will be inlined, resulting in
+
+```yul
+{
+	sstore(0, 1)
+	if calldataload(4) { return(0,0) }
```

Comment:
Is this really an optimized version of `if iszero(lt(calldatasize(), 4)) { return(0,0) }`?

Reply:
Of course it is :-P. (fixing...)